Feature: Tuist Plugins

luispadron opened the pull request:
Resolves https://github.com/tuist/tuist/issues/2021

## Short description 📝

Originally this PR intended to add the ability to customize `ProjectDescriptionHelpers` location as defined in RFC #2021. However, it makes more sense to extend this further to include generic `Plugin` support. Below is high level overview of how the API will be extended to support this and some changes required to the project generator.

## Plugin

A `Plugin` is defined as a directory that contains a `Plugin.swift` manifest file. This `Plugin.swift` must define a `plugin` that may look something like this:

```swift
// Plugin.swift

import ProjectDescription
// ⚠️ Cant import any plugins or ProjectDescriptionHelpers

let plugin = Plugin.helper(name: "MyHelperPlugin")
```

The list of supported `Plugin`s includes:

- `helper`: A plugin intended to extend `ProjectDescription`
  - These are independently compilable Swift frameworks i.e. they may only depend on `ProjectDescription`. 
  - Cannot be used in `Plugin.swift` or `Plugins.swift`.

### Implementation

`Plugin`s will be loaded before other manifests so that they may be used in other manifests such as `Project.swit`. They may then be injected and used for project generation or as in the use case of helpers, as a means of creating importable frameworks for usage within Tuist.

Because of issues that would be created by allowing this, importing `ProjectDescriptionHelpers` inside of `Plugin.swift` is disallowed. One such issue would be that becuase `ProjectDescriptionHelpers` can import plugins we would be creating a circular dependency between these two.

#### Future

- Shared templates for `tuist scaffold` (I'd love to see this, if I can find the time for it, I might try this myself) @fortmarek
- Defining custom commands that tuist can trigger - similar to when you run custom actions via Fastlane (although not sure we'd want this as it raises complexity) @fortmarek
- We could also add a feature to add custom Xcode templates that could be shared with your team (we use fastlane for this [right now](https://github.com/AckeeCZ/ios-templates)) @fortmare

## Plugins

In order to include and load a `Plugin.swift` manifest we must first define where to locate these / include these in the project.

```swift
/// Tuist/Plugins.swift

import ProjectDescription
// ⚠️ Cant import any plugins or ProjectDescriptionHelpers

let plugins = Plugins([
  .local(path: "/som/path/to/Plugin.swift"),
  .git(url: "some/repo.git", branch: "main")
])
```

`Plugins.swift` manifest defines the `Plugin` dependencies for the entire Tuist project, similar to how `Config.swift` defines project wide settings. Each of these are locations that point to a directory containing a `Plugin.swift` file.

### Implementation

We will load `Plugins.swift` manifest before we can load each `Plugin.swift` manifest since we first need to get the locations defined in `Plugins.swift` and resolve them either by cloning into the `~/.tuist/Cache` for `.git` plugins or simply resolving the path defined for `.local` plugins. Once we have resolved the `Plugins` we will iterate through each of them and load each individual `Plugin.swift` manifest. These loaded manifest can then be used as defined in `Plugin`.

Because of issues that would be created by allowing this, importing plugins or `ProjectDescriptionHelpers` inside of `Plugins.swift` is disallowed. Because we do not yet have the plugins loaded and because `ProjectDescriptionHelpers` can include plugins we cannot allow importing plugins or `ProjectDescriptionHelpers` in `Plugins.swift`.

## Example integration

### Helper plugin

We can first define a helper plugin. This is a directory that includes a `Plugin.swift` file, it may look something like this:

```bash
.
├── Plugin.swift
├── SomeOtherFile.swift
├── Helpers
│   └── ProjectHelpers.swift
└── Util
    └── System.swift
```

The `ProjectHelpers.swift` would be collected and built as defined in `Plugin.swift`. Notice that only the files under `Helpers` are used for the `.helpers` plugin in this example. This directory could then be uploaded to and version-controlled on GitHub for example under a URL like: `https://github.com/tuist/ExampleTuistHelpersPlugin`.

Our `Plugin.swift` may look something like:

```swift
// Plugin.swift

import ProjectDescription

let plugin = Plugin.helper(name: "ExampleTuistHelpers")
```

### Tuist project

In another project, we can then include the plugin by adding it to our `Tuist/Config.swift` file. It may look something like this:

```bash
.
├── Project.swift
└── Tuist
    └── Config.swift
```

Our `Config.swift` may look something like:

```swift
// Tuist/Config.swift
import ProjectDescription

let config = Config(
  plugins: [
    .git(url: "https://github.com/tuist/ExampleTuistHelpersPlugin", branch: "main")
  ]
)
```

We could then import `ExampleTuistHelpers` in our `Project.swift` for example:

```swift
// Project.swift

import ProjectDescription
import ExampleTuistHelpers

let project = Project.exampleHelperExtension(name: "App")
```

luispadron commented on Sources/TuistLoader/Mappers/ProjectDescriptionHelpersLocationMapper.swift:
```diff
@@ -0,0 +1,127 @@
+
+        let helpersGitDirectory = derivedHelpersDirectory.appending(.init(".git"))
+
+        let deleteHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .absent)
+        let createHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .present)
+        let gitClone = CommandDescriptor(command: ["git", "clone", url, derivedHelpersDirectory.pathString])
+        let gitCheckout = CommandDescriptor(command: [
+            "git",
+            "--git-dir",
+            helpersGitDirectory.pathString,
+            "--work-tree",
+            derivedHelpersDirectory.pathString,
+            "checkout",
+            gitId
+        ])
+        let removeGit = DirectoryDescriptor(path: helpersGitDirectory, state: .absent)
+        
+        return [
+            .directory(deleteHelpersDirectory),
+            .directory(createHelpersDirectory),
+            .command(gitClone),
+            .command(gitCheckout),
+            .directory(removeGit)
+        ]
+    }
```
Hey @fortmarek How did you envision using `GitController` here? From my understanding we're supposed to return `SideEffectDescriptors` in the mappers instead of explicitly performing these operations. Let me know if I'm missing something!

fortmarek replied:
Nope, this is the correct solution -> I have not realized that the git commands are going to be used in `Mapper` context 🙂

luispadron commented:
@fortmarek  @pepibumur Hey all, this is ready for an initial review. Let me know what you think / any improvements.

fortmarek commented on Sources/ProjectDescription/HelpersLocation.swift:
```diff
@@ -0,0 +1,119 @@
+        case gitWithTag(url: String, tag: String)
+        case gitWithHash(url: String, hash: String)
+
+        enum CodingKeys: CodingKey  {
```
Is this necessary? I think we can delete this

luispadron replied:
So this is needed to allow `LocationType` to conform to `Equatable`

fortmarek commented on Sources/TuistLoader/Mappers/ProjectDescriptionHelpersLocationMapper.swift:
```diff
@@ -0,0 +1,127 @@
+    private let config: TuistCore.Config
+    private let rootDirectoryLocator: RootDirectoryLocating
+
+    private lazy var derivedHelpersDirectory: AbsolutePath? = {
```
I think this can be turned into a function `private func derivedHelpersDirectory() throws -> AbsolutePath?`

That way you:
- leverage error handling that is already implemented up the chain
- do not repeat `logger.error` call

To reuse the error message, I'd create a new error that you can throw

This also ensures that the whole pipeline fails when the derived directory is not found -> that is something we want as otherwise we get into unknown state where we can not ensure the right results.

Besides that, I'd also create a test for this error being thrown.

luispadron replied:
This makes sense, updating!

fortmarek commented on Sources/TuistLoader/Mappers/ProjectDescriptionHelpersLocationMapper.swift:
```diff
@@ -0,0 +1,127 @@
+
+    // MARK: - ProjectMapping
+
+    public func map(project: TuistCore.Project) throws -> (TuistCore.Project, [SideEffectDescriptor]) {
```
Do we need to prepend `Project` with `TuistCore`? 🤔  (this appears to be in multiple places here)

luispadron replied:
Not needed, removing

fortmarek commented on Sources/TuistLoader/Mappers/ProjectDescriptionHelpersLocationMapper.swift:
```diff
@@ -0,0 +1,127 @@
+        let sourceNames = sources.map(\.basename)
+        let createHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .present)
+
+        let createFiles = zip(sourceNames, sourceContents)
```
`zip` works, but maybe we can create a new `FileCopyDescriptor` that will do the copying instead? Copying is a common operation, so it'd be great if we can reuse that logic in the future.

luispadron replied:
Agree, added!

fortmarek commented on Sources/TuistLoader/Mappers/ProjectDescriptionHelpersLocationMapper.swift:
```diff
@@ -0,0 +1,127 @@
+
+        let deleteHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .absent)
+        let createHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .present)
+        let gitClone = CommandDescriptor(command: ["git", "clone", url, derivedHelpersDirectory.pathString])
```
It is a little bit unfortunate that this clones the repository every time user generates a project. I think that maybe we can save the latest `gitId` and `url` into `Derived` directory and if its contents match to the one supplied to this function, we would not clone it at all? What do you think?

luispadron replied:
For storing the previously pulled git repo how should we serialize this info? Should we just encode the HelpersLocation struct to a file in Derived and then when performing the mapping decode it and if they match just skip downloading?

fortmarek replied:
yeah, I think that's a good approach 👍

fortmarek commented on Tests/TuistLoaderTests/Mappers/ProjectDescriptionHelpersLocationMapperTests.swift:
```diff
@@ -0,0 +1,154 @@
+    override func setUp() {
+        super.setUp()
+        rootDirectoryLocator = MockRootDirectoryLocator()
+        rootDirectoryLocator.locateStub = fakeRoot
```
how about we generate `fakeRoot` with `temporaryPath()` function? And change `fakeRoot` to `private var fakeRoot: AbsolutePath!`

fortmarek commented on Sources/ProjectDescription/HelpersLocation.swift:
```diff
@@ -0,0 +1,119 @@
+            case gitWithHash
+        }
+
+        public func encode(to encoder: Encoder) throws {
```
I'd test the custom en/decoding logic with `XCTAssertCodable` and create `HelpersLocationTests` for this

fortmarek commented on Sources/ProjectDescription/Config.swift:
```diff
@@ -52,6 +57,7 @@ public struct Config: Codable, Equatable {
 extension Config.GenerationOptions {
     enum CodingKeys: String, CodingKey {
         case xcodeProjectName
+        case helpers
```
I don't see this key being used anywhere? I'd also add it to `ConfigTests`

fortmarek reviewed (commented):
The code is in a really great shape 👏  I've added some smaller notes - could you also create a fixture test for this? Read more about our acceptance tests [here](https://tuist.io/docs/contribution/testing-strategy/#acceptance-tests). I think it'd be best to test both local and remote option - I'll add you to the tuist organization, so you can create a repository for the example project description helpers.

Besides that, could you also add a mention of this feature in the documentation?

I have not tested this locally, yet, I'll do that once the fixture is ready 🙂

luispadron commented:
@fortmarek I believe I've addressed most of your comments. 

Some open questions:

- For storing the previously pulled git repo how should we serialize this info? Should we just encode the `HelpersLocation` struct to a file in `Derived` and then when performing the mapping decode it and if they match just skip downloading?
- As soon as I have access to the org I'll add a test repo/fixture for pulling from git
   - Should this test only one of the routes (`branch`, `tag`, `hash`) or should I make 3 fixtures?

luispadron commented:
Also, one of your original asks was to lock the copied files. 

Lookin at https://github.com/cocoapods/cocoapods/issues/1154 there was some debate about whether or not to do this for cocoapods. Not sure how we all feel about this, but something we can add if we want.

pepibumur commented on Sources/ProjectDescription/HelpersLocation.swift:
````diff
@@ -0,0 +1,119 @@
+    /// ```
+    /// .local(path: "/User/local/bin")
+    /// ```
+    public static func local(path: Path) -> Self {
````
Tuist aims for convention over configuration and I think this is a good example where the convention is useful. I think the opinion on where helpers should leave should be Tuist's. The more decisions that Tuist makes the fewer users have to make and thus focus on the important things. Having said that, I'd turn this into:

```swift
public var local: Self {}
```

fortmarek replied:
This is a location to where your reused location helpers are, not where they should be saved, so I believe it should stay as-is

luispadron replied:
Correct, this was intended so that you could store helpers somewhere on your machine locally. This doesn't determine where they will be copied to.

luispadron replied:
Having said that however, if we dont want to maintain the code required for local helpers/plugins and would rather just default to remote that works for me as well.

fortmarek replied:
I think it is useful during development of the remote helpers to be able to have a local version (instead of having to constantly push your code), so I'm in favor of keeping it

pepibumur reviewed (commented):
👋 Hi @luispadron. First of all, thanks for taking this project. I think it's going to be very useful for teams that are building white-label apps and need to share helpers across repositories.

Before I review further, I think we should revisit the solution proposed in this PR (sorry I couldn't catch this earlier). 

- Mappers are intended to be used for modifying the graph and its models. Along the mapping process, mappers can generate side effects that have a connection with the generated project (e.g. generate Info.plist files). For that reason, I don't think mappers are suitable to model this feature.
- Moreover, I think the remote helpers should be placed and cached in Tuist's global directory *(`~/.tuist/...`)* uniquely identified with a fingerprint that is generated from the Git URL and the branch/commit/sha. I think remote helpers should have a name, and that will be the name of the generated module that you need to use when importing them: `import MyNameProjectDescriptionHelpers`.

The fetching and the compilation of the helpers into a module should happen early in the process. We can compare it to the Swift Package Manager where all the third-party dependencies are pulled and compiled, and then SPM's build system builds your package.

Since I anticipate extending Tuist with remote code as we are doing with helpers in this case, what about making this implementation more generic and call it **plugins.**

A plugin is a repository that contains a `Plugin.swift` manifest file with the following content:

```swift
import ProjectDescription

let plugin = Plugin(name: "MyTuistPlugin", helpers: true)
```

Then your project's `Config.swift` can have something along the lines of:

```swift
// Config.swift
import ProjectDescription

let config = Config(plugins: [.git("https://github.com/tuist/my-plugin", sha: "xyz")])
```

Then the generation logic would clone, hash, and store the plugin in a global directory: `~/.tuist/Cache/Plugins` and we can extend the manifest loader logic to build and expose the helpers that are defined in plugins too :). 

I know I'm proposing a lot of changes to your original proposal, but since you are the first one taking the steps towards extending Tuist with "something" coming from a remote repository, why not taking the opportunity to create a foundation for more future improvements?

Let me know if you want to pair on kicking this off. I'll be happy to help

fortmarek commented:
Thanks for the input @pepibumur and for bringing the bigger picture here.

I have a few notes:

I'd convert 
```swift
let plugin = Plugin(name: "MyTuistPlugin", helpers: true)
```
to be enum:
```swift
let plugin = Plugin.helpers(name: "MyTuistPlugin")
```

This will be then more easily extendable for different types of plugins in the future.

Bundling the remote helpers into their own library is a really good idea 💡 - this also means that we won't need to lock the files as the user will not have direct access to the files, @luispadron.

> Along the mapping process, mappers can generate side effects that have a connection with the generated project

I have not thought about it this way, but I suppose it makes sense not to use mappers for this (there is a slight connection with the generated project, but not a strong one, indeed). Would you do it then via interactor in pre generation actions? (similar to `SigningInteractor`).

For @luispadron questions:

> As soon as I have access to the org I'll add a test repo/fixture for pulling from git 

You should have the access now!

>  Should this test only one of the routes (branch, tag, hash) or should I make 3 fixtures?

I think it's fine to make one fixture only.

luispadron commented:
@pepibumur @fortmarek Thanks for the input! I agree that if we're going to add something like this to make it generic as possible so we can extend it to more than just helpers. 

I'll need some help further understanding how to integrate these additional requirements. Specifically, how plugins will work and how to pull them into a module, as well as performing fingerprinting on the files. 

I have some extra time this week to work on this. We can either setup sometime or just async it via comments here so things are documented.

fortmarek commented:
I am not sure what's the schedule of @pepibumur but if he's busy I can help you out, maybe tomorrow, if you want to, @luispadron

> how plugins will work and how to pull them into a module

First, you'll need to clone the repository specified in `Config.swift` (taking from @pepibumur 's example) to `~/.tuist/Cache/Plugins`. Once that is done you'll then have to load the plugin manifest (you can probably implement this to `ManifestLoader.swift` and take inspiration from loading other manifests, such as `Config.swift` - note that the logic for `Plugin.swift` will be a little bit different as you don't want to load `ProjectDescriptionHelpers` which is something we do for eg. `Config.swift`). With the manifest loaded, you can then derive what your next steps should be depending on the plugin - if it's a helper, you'll have to build a module. For that you can use `ProjectDescriptionHelpersBuilder` - I think you can just modify its input parameters and it should successfully build the helpers that reside in the tuist cache (or locally at predefined path if the helpers are local). 

Then it depends if you are doing `tuist edit` or a different command such as `tuist generate`. For `tuist edit` you'll have to import the remote project description helpers in `ProjectMapperEditor.swift` and if `tuist generate`, you'll have to import them in `ManifestLoader.swift` in `func loadManifestData(at path: AbsolutePath)` for it to work properly.

> performing fingerprinting on the files

For that you can simply use `ContentHasher` 

Those are a few tidbits, but there is definitely more to the implementation. 

I'd also recommend creating the fixture first and working with that, it has usually made it easier for me to test my current solution.

luispadron commented:
This is super helpful! I'll give it a try and if I end up needing some more help we can sync up.

luispadron commented:
@fortmarek @pepibumur So I've gotten this to a point where we can define a `Plugin.swift` manifest and load it from a file. I need some suggestions in terms of how to use these loaded plugins to perform what we need.

Would we just need to now pass these `Plugin`s to the graph when it goes to load any of the other manifests and build it in the same way that `ProjectDescriptionHelpersBuilder` does?

fortmarek commented:
Hey @luispadron !

Looking briefly at your code, these are some suggestions I thought of (but you have now more context, so feel free not to take them!):
- I'd refactor `PluginService` into its own framework. We follow [microfeatures architecture](https://tuist.io/docs/building-at-scale/microfeatures/) and adding support for plugins is something that makes sense to add as a new feature. 
- Do not be bound by `Plugin.swift` and what you write it in there -> the way you structure the public API can be quite different from what you do internally. I think `Plugin` model can stay in the public API and you can pluck out the things you need. So, if you parse a `Plugin` of type `.helpers`, you can turn it into a new model:
```swift
struct CustomProjectDescriptionHelpers { 
  /// Location of where they are located
  let path: AbsolutePath  
  
  /// Name of the module
  let name: String
}
```

This model would then be easier to pass around. I see that you have a function `loadPlugin() -> Plugin` -> maybe we can return `Plugins` that will look like this:
```swift
struct Plugins {
  let customProjectDescriptionHelpers: [CustomProjectDescriptionHelpers]
}
```

In `Generator.swift` you can then easily access `customProjectDescriptionHelpers` property and pass *that* to `loadWorkspace` / `loadProject`. You'll then need to pass it all the way to where you parse manifests in `ManifestLoader` and compile those custom helpers if there are any.

Passing this property is just one way how to approach this.

If that turns out to be too hard and necessitates too many changes, you can leverage the fact that the helpers are already at a known location somewhere in `~/.tuist/Cache`. If you make a folder there for each project (which I am already working in a different PR), you could just parse the list of directories that should include project description helpers and add them automatically -> the benefit here is that this can be done right in `ManifestLoader` without passing of any properties (but it's not as safe, so I'm in favor of the first approach).

> Would we just need to now pass these Plugins to the graph when it goes to load any of the other manifests and build it in the same way that ProjectDescriptionHelpersBuilder does?

To answer your question more concretely: yes, we need to somehow pass plugins (or their representation to the other manifests) and we should build it in the same way as `ProjectDescriptionHelpersBuilder` does - I'd try to reuse the logic of this component.

Again, thanks for doing great progress on this and do take this advice but I have poked in the code just briefly, so I might have missed something that makes my advice useless 😛

luispadron commented:
Update: Progress continues on this. I am able to generate projects using helpers and link the plugin modules when creating the Project. I'll be looking more at this this week and adding documentation and tests to get the PR into a mergeable state. Thanks again for all the help on getting this ready, its been super helpful and good to see how dedicate you all are to the project. In the meantime feel free to review some more of the code and leave feedback before things get too solidified. 

One change: I changed from the original discussion is that I don't feel the need to append `ProjectDescriptionHelpers` to any helper plugins. This seems counter-intuitve to the `Plugin.helpers(name: "")` API and I would expect my plugin to be built with the name I supplied (without additions). 

One ask: What other contexts do we see plugins being used for in the future? I'd like to get an idea for when I go write some documentation for this feature.

@fortmarek @pepibumur

fortmarek commented:
Hi @luispadron, thanks for the update!

> I changed from the original discussion is that I don't feel the need to append ProjectDescriptionHelpers to any helper plugins. This seems counter-intuitve to the Plugin.helpers(name: "") API and I would expect my plugin to be built with the name I supplied (without additions).

I agree, if users want to append ProjectDescriptionHelpers, they are free to do so )

> What other contexts do we see plugins being used for in the future? I'd like to get an idea for when I go write some documentation for this feature.

This is an interesting question! 
So far I came up with this, but there are definitely more opportunities to use the new plugin system:
- Shared templates for `tuist scaffold` (I'd love to see this, if I can find the time for it, I might try this myself)
- Defining custom commands that tuist can trigger - similar to when you run custom actions via Fastlane (although not sure we'd want this as it raises complexity)
- we could also add a feature to add custom Xcode templates that could be shared with your team (we use fastlane for this [right now](https://github.com/AckeeCZ/ios-templates))

Let me know when you want me to look over the code / try out the feature!

fortmarek commented on Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift:
```diff
@@ -41,52 +47,89 @@ final class ProjectDescriptionHelpersBuilder: ProjectDescriptionHelpersBuilding
     }
 
-    // MARK: - Fileprivate
+    private func createCommand(
```
I am wondering (I hope this is a good part where to ask that) - are you able to import custom project description helpers in your default ones? Eg. there is a nifty method in a remote helper, but I'd like to build on it in my own `ProjectDescriptionHelpers`, is that possible? From what I could gather, I don't see it, but maybe I am missing something. I believe this is gonna be quite a common scenario, so it'd be great to support this from the start.

I'd also add/modify a fixture to catch this scenario.

luispadron replied:
You are correct that as it stands we cannot import plugins from the default helpers. I originally did not add this support because this way we don't need to worry about cyclic dependency between Plugins and `ProjectDescriptionHelpers`. 

However, if we want to be able to import plugin helpers into the "default" helpers we will need to remove the ability to import the "default" helpers in plugins. I think this makes sense, so I'll be doing that now.

fortmarek replied:
Yes, I think it makes sense not to need `ProjectDescriptionHelpers` in `Plugin.swift` 👍

fortmarek commented on Sources/TuistPlugin/Service/PluginService.swift:
```diff
@@ -0,0 +1,101 @@
+}
+
+/// A default implementation of `PluginServicing`.
+public class PluginService: PluginServicing {
```
I'd mark this `final`

fortmarek commented on Sources/TuistPlugin/Service/PluginService.swift:
```diff
@@ -0,0 +1,101 @@
+
+        var projectDescriptionHelpers = [CustomProjectDescriptionHelpers]()
+
+        zip(pluginManifests, pluginPaths).forEach { plugin, path in
```
I think this can be further simplified into something like this:
```swift
let projectDescriptionHelpers = zip(pluginManifests, pluginPaths)
.map { ... return .init(name: name, path: path) }
```

`projectDescriptionHelpers` will then no longer have to mutable 🙂

fortmarek commented on Sources/TuistPlugin/Service/PluginService.swift:
```diff
@@ -0,0 +1,101 @@
+    /// the Tuist cache with a unique fingerprint
+    private func fetchGitPlugin(at url: String, with gitId: String, fingerprint: String) throws -> AbsolutePath {
+        let pluginDirectory = Environment.shared.cacheDirectory
+            .appending(.init(Constants.PluginDirectory.name))
```
I would not use `.init` as it's a little bit hard to read and derive the type just by reading this (I generally try to keep usage of `.init` low as it makes the code a little bit less explicit)

fortmarek commented on Sources/TuistSupportTesting/Utils/MockGitHandler.swift:
```diff
@@ -0,0 +1,27 @@
+    
+    public private(set) var cloneInCallCount = 0
+    public private(set) var cloneInParameters: [(url: String, path: AbsolutePath)] = []
+    public func clone(url: String, in path: AbsolutePath) throws {
```
I personally think the approach of using a closure here is usable in more contexts and it might be cumbersome, in some cases, to use it like this

by closure I mean stubbing the whole method in the following way:
```swift
public var cloneStub: ((String, AbsolutePath) throws -> ())?
```

But I know that this style is already in the codebase, so stick to whatever *you* prefer 🙂

fortmarek commented on Tests/ProjectDescriptionTests/PluginLocationTests.swift:
```diff
@@ -0,0 +1,24 @@
+
+    func test_codable_gitWithBranch() throws {
+        let subject = PluginLocation.git(url: "https://git.com/repo.git", branch: "main")
+        XCTAssertCodable(subject)
```
I'd also add plugins to `ConfigTests`

fortmarek commented on Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift:
```diff
@@ -38,13 +38,13 @@ final class ProjectDescriptionHelpersBuilderIntegrationTests: TuistTestCase {
         XCTAssertNotNil(FileHandler.shared.glob(path, glob: "*/*/libProjectDescriptionHelpers.dylib").first)
         XCTAssertNotNil(FileHandler.shared.glob(path, glob: "*/*/ProjectDescriptionHelpers.swiftdoc").first)
-        XCTAssertTrue(FileHandler.shared.exists(paths.first!!))
+        XCTAssertTrue(FileHandler.shared.exists(paths[0][0].path))
```
indexing is basically force unwrapping, so this code is not much safer (although maybe a little bit readable?) - I'd use `XCTUnwrap` here

fortmarek reviewed (commented):
I think the code is really solid, well done 👏 This is gonna be a great feature and I really like the code. I have not found anything wrong structurally and it captures well the results of our PR/RFC discussion - just added some comments, but most of them are nitpicks 🙂

luispadron commented on Sources/TuistCore/Graph/GraphLoader.swift:
```diff
@@ -118,29 +149,38 @@ public class GraphLoader: GraphLoading {
         }
     }
 
+    public func loadPlugin(path: AbsolutePath) throws -> Plugin {
```
```swift
let pluginManifests = try pluginPaths.map(graphLoader.loadPlugin)
```

luispadron replied:
Actually, maybe i should remove it from `GraphLoading` and only do this in `ManfiestLoader`

luispadron commented on Sources/TuistGenerator/GraphViz/GraphVizGenerator.swift:
```diff
@@ -2,6 +2,7 @@ import Foundation
 import GraphViz
 import TSCBasic
 import TuistCore
+import TuistPlugin
```
That makes sense, ive mode this to TuistKit and removed `TuistPlugin` dependency from `TuistGenerator`. This is now injected.

luispadron commented on Sources/TuistLoader/Loaders/CachedManifestLoader.swift:
```diff
@@ -1,6 +1,7 @@
 import Foundation
 import ProjectDescription
 import TSCBasic
+import TuistCore
```
Hmm I get `No such module TusitCore.Plugins` when i do this

luispadron commented on Sources/TuistLoader/Loaders/ManifestLoader.swift:
```diff
@@ -189,10 +206,16 @@ public class ManifestLoader: ManifestLoading {
 
-    private func loadManifestData(at path: AbsolutePath) throws -> Data {
+    private func loadManifestData(
+        for manifest: Manifest,
```
This was add when i was excluding including project description helpers for plugins, not needed anymore

luispadron commented:
@fortmarek @pepibumur 

 Trying to add support for plugin import in default helpers but have come to a circular dependency issue. 

The problem:
- As it stands today all manifests have the ability to import the default `ProjectDescriptionHelpers`.
- We would like the default `ProjectDescriptionHelpers` to be able to import any helper plugins, since ideally any file that can import `ProjectDescription` should also be able to import helper plugins.
- Config can import `ProjectDescriptionHelpers`
- If `ProjectDescriptionHelpers` is imported by Config manifest and `ProjectDescriptionHelpers` depends on some helper plugins, then how can we build the plugins **before** loading the Config manifest? 

This seems like a chicken-egg problem where we need the config to build and determine the plugins but we need the plugins to be able to build the config. Been thinking about this for a day now and I can't really come up with anything solid besides disallowing imports of **any** helpers (including default helpers) in the `Config` manifest. This seems like it would be a breaking change from what we do today and it's not an elegant solution.

Do you guys have any thoughts on how to do this in such a way that we would not have this circular dependency?

fortmarek commented:
Huh, yeah, that's a good catch @luispadron! Thanks for taking your time to explain the issue in-depth.

I can't think of a better solution than those that you have laid out. 

I am more inclining to creating a new `Plugins.swift` manifest as removing ProjectDescriptionHelpers would be, as you said, breaking plus I think we should let users import the helpers there. cc @pepibumur

luispadron commented:
Proposal for `Plugins.swift`

```swift
// Plugins.swift
// ⚠️ Cant import ProjectDescriptionHelpers or plugins

let plugins = Plugins([
  .local(path: "/som/path/to/Plugin.swift"),
  .git(url: "some/repo.git", branch: "main")
])
```

Random thought: We could use function builders to make the api more declarative / extensible:

```swift
let plugins = Plugins {
  Local(path: "/som/path/to/Plugin.swift")
  Git(url: "some/repo.git", branch: "main")
}
```

First option is more standard/stable though

luispadron commented:
I've also updated the original discussion to include some more information. This will be a starting point for the documentation of Plugin feature.

fortmarek commented:
So, after getting input from @pepibumur, we think that it's fine to make a breaking change. There is not much of a need to add helpers to `Config.swift` as they are mainly meant to be used across Project manifests. 

This can help us going further as it's easier to maintain less manifest file types - and it's also more comprehensible for the users.

That means you can simply delete importing of helpers to `Config.swift` and that should resolve the circular dependency 🙂 

We'll just need to make sure that we communicate this breaking change properly, so users know why this change is being made.

Also, thanks @luispadron for updating the PR notes with the current state of how the feature should work 👏 

Reading through it, I have just one note:

> The SomeHelpers.swift, ProjectHelpers.swift and System.swift would be collected and built as defined in Plugin.swift

I think it might be better to let users use only a single directory, eg. `Helpers`, from which we would take all swift files. I would not expect the Swift files in root directory to be used.

In the future, it'd also be great if you could edit the plugin and its helper files via `tuist edit` (I believe that's not possible right now, correct me if I'm wrong) - but that can be left for the next iteration.

luispadron commented:
Awesome! Glad we're going the route of using Config as adding yet another manifest was sounding like a problem. I'll take these comments in mind as I continue to work on this feature.

pepibumur reviewed (commented):
Hey @luispadron would you mind splitting this one into smaller pieces? I think that'll help with making reviews more effective and we'll be able to catch issues that otherwise will go unnoticed. I propose the following PRs:
- ProjectDescription models.
- TuistCore models and mapping logic.
- Command and service _(without the business logic)_
- Business logic.

luispadron commented:
@pepibumur Yep! I'll go ahead and split this up. Just want to make sure we have integrated it fully first then we can just split it up as needed. I'll open up separate PRs as soon as we get some more clarity on the final API

luispadron commented:
Closing in favor of #2095 and PRs which will follow it.
